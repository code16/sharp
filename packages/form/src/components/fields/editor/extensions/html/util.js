import { DOMParser } from "@tiptap/pm/model";

/**
 * @returns HTMLElement
 */
export function elementFromString(value) {
    const wrappedValue = `<body>${value}</body>`
    return new window.DOMParser().parseFromString(wrappedValue, 'text/html').body
}


function prosemirrorParse(node, schema) {
    const parser = DOMParser.fromSchema(schema);
    const dom = elementFromString(node.outerHTML);
    return parser.parseSlice(dom).content;
}

export function setupContent(content, schema) {
    const dom = elementFromString(content);
    return setupContentDOM(dom, schema);
}

export function setupContentDOM(dom, schema) {
    [...dom.children].forEach(node => {
        // in some cases we want to extract the html content if known (e.g. embeds)
        if(node.hasAttribute('data-html-content') && node.children.length === 1) {
            const childFragment = prosemirrorParse(node.firstElementChild, schema);
            if(childFragment.firstChild?.type.name.startsWith('embed:')) {
                node.parentElement.replaceChild(node.firstElementChild, node);
            }
            return;
        }

        const parsed = prosemirrorParse(node, schema);
        if(!parsed.size) {
            const div = document.createElement('div');
            div.setAttribute('data-html-content', 'true');
            node.parentElement.replaceChild(div, node);
            div.appendChild(node);
        }
    });
    return dom.innerHTML;
}
