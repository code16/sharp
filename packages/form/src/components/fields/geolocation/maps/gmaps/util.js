

export const defaultMapOptions = {
    mapTypeControl: false,
    streetViewControl: false
};

export function normalizeBounds(latLngBounds) {
    return [
        latLngBounds.getSouthWest().toJSON(),
        latLngBounds.getNorthEast().toJSON(),
    ];
}

export function toLatLngBounds(normalizedBounds) {
    const bounds = normalizedBounds;
    return Array.isArray(bounds)
        ? new google.maps.LatLngBounds(bounds[0], bounds[1])
        : null;
}

export function createMapOptions({ maxBounds, ...options }) {
    const res = { ...options };
    if(Array.isArray(maxBounds)) {
        res.restriction = {
            latLngBounds: toLatLngBounds(maxBounds),
        };
    }
    return res;
}