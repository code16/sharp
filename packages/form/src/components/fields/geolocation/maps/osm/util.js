import { latLngBounds } from 'leaflet';

export function normalizeBounds(latLngBounds) {
    const sw = latLngBounds.getSouthWest();
    const ne = latLngBounds.getNorthEast();
    return [
        { lat: sw.lat, lng: sw.lng },
        { lat: ne.lat, lng: sw.lng },
    ];
}

export function toLatLngBounds(normalizedBounds) {
    const bounds = normalizedBounds;
    return Array.isArray(bounds)
        ? latLngBounds(bounds[0], bounds[1])
        : null;
}