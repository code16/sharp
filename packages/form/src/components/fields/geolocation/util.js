export function dd2dms(D, lng){
    let dir = D<0?lng?'W':'S':lng?'E':'N',
        deg = 0|(D<0?D=-D:D),
        min = 0|D%1*60,
        sec = (0|D*60%1*6000)/100;

    return `${deg}°${min}"${sec}' ${dir}`;
}


export function providerName(providerData) {
    return providerData
        ? providerData.name
        : null;
}

export function providerOptions(providerData) {
    return providerData
        ? providerData.options || {}
        : {};
}

export function tilesUrl(mapsOptions) {
    return mapsOptions
        ? mapsOptions.tilesUrl
        : null;
}

export function triggerResize() {
    if (typeof Event === 'function') {
        window.dispatchEvent(new Event('resize'));
    } else {
        const evt = document.createEvent('UIEvents');
        evt.initUIEvent('resize', true, false, window, 0);
        window.dispatchEvent(evt);
    }
}