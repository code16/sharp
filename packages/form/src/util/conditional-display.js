import { logError } from 'sharp';

const helpers = {
    computeSelectCondition({ condValues, fieldValue, isSingleSelect }) {
        if (Array.isArray(condValues)) {
            return isSingleSelect
                ? condValues.some(val => val == fieldValue)
                : condValues.some(cval => fieldValue.some(fval => fval == cval));
        }
        // 'values' is a string
        if (condValues[0] === '!') {
            let condVal = condValues.substring(1);
            return isSingleSelect
                ? condVal != fieldValue
                : !fieldValue.some(fval => fval == condVal);
        }
        // 'values' is not negative
        return isSingleSelect ? condValues == fieldValue : fieldValue.some(fval => fval == condValues);
    }
};

export function computeCondition(fields, data, condition) {
    let res = true;

    let { operator } = condition;

    if(operator !== 'or' && operator !== 'and') {
        logError(`Conditional display : unknown operator '${operator}'`, condition);
        return true;
    }

    for(let condField of condition.fields) {
        if(!(condField.key in fields)) {
            logError(`Conditional display : can't find a field with key '${condition.key}' in 'fields'`, condition);
            res = true;
        }

        let field = fields[condField.key];
        let value = data[condField.key];


        if(field.type === 'autocomplete' || field.type === 'select' || field.type === 'tags') {
            res = helpers.computeSelectCondition({
                condValues: condField.values,
                fieldValue: field.type === 'autocomplete' ? value && value.id : value,
                isSingleSelect: field.type === 'select' && !field.multiple || field.type === 'autocomplete'
            });
        }
        else if(field.type === 'check') {
            if(typeof condField.values !== "boolean") {
                logError(`Conditional display : 'values' must be a boolean for a 'check' field ('${condField.key}')`,condition,field);
                res = true;
            }
            else res = (!!value === condField.values);
        }
        else {
            logError(`Conditional display : unprocessable field type '${field.type}'`, field);
            res = true;
        }

        if(operator==='and' && !res)
            return false;
        if(operator==='or' && res)
            return true;
    }
    return res;
}

export {
    helpers
}