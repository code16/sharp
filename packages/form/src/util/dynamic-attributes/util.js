import get from 'lodash/get';
import buildTemplate from 'lodash/template';
import { logError } from "sharp";

const interpolateRE = /{{([\s\S]+?)}}/g;

export function getEmptyValueSources({ contextSources, contextData }) {
    return contextSources.filter(sourceKey => {
        const value = contextData[sourceKey];
        return value == null || value === '';
    });
}

export function warnEmptyValue(emptySources, { sourceName, attributeName }) {
    emptySources.forEach(sourceKey => {
        logError(`Dynamic attribute '${attributeName}' of the field '${sourceName}' has invalid source:
                  The field '${sourceKey}' has value null, undefined or empty string.\n 
                  Ensure the field not clearable and has initial default value.\n
                  You can also define a default value for the attribute.`);
    });
}

export function getDynamicAttributeOptions(dynamicAttributes, attributeName) {
    return (dynamicAttributes || [])
        .find(option => option.name === attributeName);
}

export function getValueFromMap({ map, path, contextData }) {
    const valuePath = path.map(key => contextData[key]);
    return get(map, valuePath);
}

export function getSourcesFromTemplate(template) {
    return [...template.matchAll(interpolateRE)].map(match => match[1].trim());
}

export function getValueFromTemplate({ template, sources, contextData }) {
    const compile = buildTemplate(template, {
        interpolate: interpolateRE,
        evaluate: false,
        escape: false,
    });
    const defaultData = (sources||[]).reduce((res, key) => ({ ...res, [key]: null, }), {});
    return compile({
        ...defaultData,
        ...contextData,
    });
}