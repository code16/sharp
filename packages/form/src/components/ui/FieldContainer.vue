<template>
    <div class="SharpFieldContainer SharpForm__form-item" :class="formGroupClasses" :style="extraStyle">
        <div class="row">
            <div class="col">
                <label v-if="showLabel" class="SharpForm__label" @click="triggerFocus">
                    {{label}}
                </label>
            </div>
            <template v-if="fieldProps.localized">
                <div class="col-auto">
                    <FieldLocaleSelect
                        :locales="$form.locales"
                        :current-locale="locale"
                        :field-value="resolvedOriginalValue"
                        :is-locale-object="isLocaleObject"
                        @change="handleLocaleChanged"
                    />
                </div>
            </template>
        </div>
        <Field v-bind="exposedProps"
            @error="setError"
            @ok="setOk"
            @clear="clear"
            @blur="handleBlur"
            ref="field"
        />
        <div class="SharpForm__form-requirement">{{stateMessage}}</div>
        <small class="SharpForm__help-message">{{helpMessage}}</small>
    </div>
</template>

<script>
    import { logError } from 'sharp';
    import { ErrorNode, ConfigNode }  from 'sharp/mixins';
    import Field from '../Field';
    import FieldLocaleSelect from './FieldLocaleSelect';
    import { resolveTextValue, isLocalizableValueField } from '../../util/locale';


    export default {
        name: 'SharpFieldContainer',

        mixins: [ ErrorNode, ConfigNode ],

        components: {
            Field,
            FieldLocaleSelect,
        },

        inject:['$tab', '$form'],

        props : {
            ...Field.props,

            label: String,
            helpMessage: String,
            originalValue: [String, Number, Boolean, Object, Array, Date],
        },
        data() {
            return {
                state:'classic',
                stateMessage:''
            }
        },
        watch: {
            value() {
                if(this.state === 'error')
                    this.clear();
            },
            '$form.errors'(errors) {
                this.updateError(errors);
            },
            locale() {
                this.updateError(this.$form.errors);
            }
        },
        computed: {
            formGroupClasses() {
                return [
                    `SharpForm__form-item--type-${this.fieldType}`,
                    {
                        'SharpForm__form-item--danger': this.state==='error',
                        'SharpForm__form-item--success': this.state==='ok',
                        'SharpForm__form-item--no-label': !this.showLabel,
                    }
                ];
            },
            extraStyle() {
                return this.fieldProps.extraStyle;
            },
            exposedProps() {
                return {
                    ...this.$props,
                    uniqueIdentifier: this.mergedErrorIdentifier,
                    fieldConfigIdentifier: this.mergedConfigIdentifier
                };
            },
            showLabel() {
                return !!this.label || this.label === '';
            },
            resolvedOriginalValue() {
                return resolveTextValue({ field:this.fieldProps, value:this.originalValue });
            },
            isLocaleObject() {
                return isLocalizableValueField(this.fieldProps);
            }
        },
        methods: {
            updateError(errors) {
                let error = errors[this.mergedErrorIdentifier];
                if(error == null) {
                    this.clear();
                }
                else if(Array.isArray(error)) {
                    this.setError(error[0]);
                }
                else {
                    logError(`FieldContainer : Not processable error "${this.mergedErrorIdentifier}" : `, error);
                }
            },
            setError(error) {
                this.state = 'error';
                this.stateMessage = error;
                if(this.$tab) {
                    this.$tab.$emit('error', this.mergedErrorIdentifier);
                }
            },
            setOk() {
                this.state = 'ok';
                this.stateMessage = '';
            },
            clear() {
                this.state = 'classic';
                this.stateMessage = '';
                if(this.$tab) {
                    this.$tab.$emit('clear', this.mergedErrorIdentifier);
                }
                this.$form.$emit('error-cleared', this.mergedErrorIdentifier);
            },
            triggerFocus() {
                this.$set(this.fieldProps,'focused',true);
            },
            handleBlur() {
                this.$set(this.fieldProps,'focused',false);
            },
            handleLocaleChanged(locale) {
                this.$emit('locale-change', this.fieldKey, locale);
            }
        },
        mounted() {
            //console.log(this);
        }
    }
</script>