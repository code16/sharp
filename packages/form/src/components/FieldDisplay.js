import { logError } from 'sharp';
import { UnknownField } from "sharp/components";
import FieldContainer from './ui/FieldContainer';
import { isLocalizableValueField } from "../util/locale";
import { computeCondition } from '../util/conditional-display';


export function acceptCondition (fields, data, condition) {
    if(!condition)
        return true;

    return computeCondition(fields,data,condition);
}

const getValue = (form, field, value, locale) => {
    if(form.localized && field.localized && value && isLocalizableValueField(field)) {
        return value[locale];
    }

    return value;
};

const getIdentifier = (identifier, field, locale) => {
    if(field.localized)
        return `${identifier}.${locale}`;
    return identifier;
};


export default {
    name: 'SharpFieldDisplay',
    functional: true,

    inject:['$form'],



    render(h, { props, injections, data }) {
        let { fieldKey,
            contextFields,
            contextData,
            errorIdentifier,
            updateVisibility,
            readOnly,
            ...sharedProps } = props;

        let { $form } = injections;

        let field = contextFields[fieldKey];
        let value = contextData[fieldKey];

        if(!(fieldKey in contextFields)) {
            logError(`Field display ('layout') : Can't find a field with key '${fieldKey}' in 'fields'`,contextFields);
            return h(UnknownField, { props: { name: fieldKey } });
        }

        let isVisible = acceptCondition(contextFields, contextData, field.conditionalDisplay);

        updateVisibility && updateVisibility(fieldKey, isVisible);

        return isVisible ? h(FieldContainer, {
            ...data,
            attrs: {
                fieldKey,
                fieldProps: {
                    ...field,
                    readOnly: readOnly || field.readOnly,
                },
                fieldType: field.type,
                value: getValue($form, field, value, props.locale),
                originalValue: value,
                label: field.label,
                helpMessage: field.helpMessage,
                errorIdentifier: getIdentifier(errorIdentifier, field, props.locale),
                ...sharedProps
            }
        }) : null

    }
}
