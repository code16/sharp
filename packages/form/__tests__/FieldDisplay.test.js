import Vue from 'vue';
import FieldContainer from '../src/components/ui/FieldContainer.vue';
import FieldDisplay from '../src/components/FieldDisplay';
import * as conditions from '../src/util/conditional-display';

import { shallowMount } from '@vue/test-utils';


jest.mock('../src/components/ui/FieldContainer.vue', () => ({
    name: 'FieldContainer',
    template: '<div class="FIELD_CONTAINER_MOCK"></div>',
}));

describe('field-display', () => {

    function createWrapper({ propsData, provide }) {
        return shallowMount(FieldDisplay, {
            context: {
                props: {
                    fieldKey: 'title',
                    configIdentifier: 'title',
                    errorIdentifier: 'title',
                    ...propsData,
                },
            },
            provide: {
                $form: new Vue({ data: { localized: false } }),
                ...provide,
            }
        });
    }

    let computeSelectCondition = ()=>{}, computeCondition = ()=>{};

    beforeEach(() => {
        computeSelectCondition = conditions.helpers.computeSelectCondition = jest.fn(conditions.helpers.computeSelectCondition);
        computeCondition = conditions.computeCondition = jest.fn(conditions.computeCondition);
    });

    test('can mount field display', () => {
        const wrapper = createWrapper({
            propsData: {
                contextData: {
                    title: null
                },
                contextFields: {
                    title: { type: 'text' }
                }
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    describe('conditional diplay', ()=>{

        test('check',  () => {
            let wrapper = null;
            const contextFields = {
                title: {
                    type: 'text',
                    conditionalDisplay: {
                        operator: 'or',
                        fields: [{ key:'check', values: true }]
                    }
                },
                check: { type: 'check' }
            };
            const contextData = {
                title: null,
                check: true
            };

            wrapper = createWrapper({
                propsData: {
                    contextData,
                    contextFields,
                }
            });
            expect(wrapper.findAll(FieldContainer).length).toBe(1);
            expect(computeCondition)
                .toHaveBeenCalledWith(contextFields, contextData, contextFields.title.conditionalDisplay);

            wrapper = createWrapper({
                propsData: {
                    contextData: {
                        ...contextData,
                        check: false,
                    },
                    contextFields,
                },
            })
            expect(wrapper.findAll(FieldContainer).length).toBe(0);
        });

        test('multiple select (unique values)', async () => {
            let wrapper = null;
            const contextFields = {
                title: {
                    type: 'text',
                    conditionalDisplay: {
                        operator: 'or',
                        fields: [
                            { key:'select', values: 2 }
                        ]
                    }
                },
                select: {
                    type: 'select',
                    multiple: true
                }
            };

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: [2, 3]
                    },
                },
            });
            expect(computeSelectCondition).toHaveBeenCalledWith({ condValues: 2, fieldValue: [2,3], isSingleSelect: false });
            expect(wrapper.findAll(FieldContainer).length).toBe(1);


            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: [3]
                    },
                }
            });
            expect(computeSelectCondition).toHaveBeenCalledWith({ condValues: 2, fieldValue: [3], isSingleSelect: false });
            expect(wrapper.findAll(FieldContainer).length).toBe(0);
        });

        test('multiple select (multiple values)', () => {
            let wrapper = null;
            const contextFields = {
                title: {
                    type: 'text',
                    conditionalDisplay: {
                        operator: 'or',
                        fields: [
                            { key:'select', values: [ 2, 6 ] }
                        ]
                    }
                },
                select: {
                    type: 'select',
                    multiple: true
                }
            }


            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: [2, 3]
                    }
                }
            });
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: [2,6], fieldValue: [2,3], isSingleSelect: false });
            expect(wrapper.findAll(FieldContainer).length).toBe(1);

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: [4, 5]
                    }
                }
            });
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: [2,6], fieldValue: [4,5], isSingleSelect: false });
            expect(wrapper.findAll(FieldContainer).length).toBe(0);
        });

        test('single select (unique values)', () => {
            let wrapper = null;
            const contextFields = {
                title: {
                    type: 'text',
                    conditionalDisplay: {
                        operator: 'or',
                        fields: [
                            { key:'select', values: 4 }
                        ]
                    }
                },
                select: {
                    type: 'select'
                }
            }

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: 4
                    }
                }
            })
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: 4, fieldValue: 4, isSingleSelect: true });
            expect(wrapper.findAll(FieldContainer).length).toBe(1);


            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: 5
                    }
                }
            })
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: 4, fieldValue: 5, isSingleSelect: true });
            expect(wrapper.findAll(FieldContainer).length).toBe(0);
        });

        test('single select (multiple values)', () => {
            let wrapper = null;
            const contextFields = {
                title: {
                    type: 'text',
                    conditionalDisplay: {
                        operator: 'or',
                        fields: [
                            { key:'select', values: [4, 8] }
                        ]
                    }
                },
                select: {
                    type: 'select'
                }
            }

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: 4,
                    }
                }
            });
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: [4,8], fieldValue: 4, isSingleSelect: true });
            expect(wrapper.findAll(FieldContainer).length).toBe(1);

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: 3,
                    }
                }
            });
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: [4,8], fieldValue: 3, isSingleSelect: true });
            expect(wrapper.findAll(FieldContainer).length).toBe(0);
        });

        test('multiple select (negative)', async () => {
            let wrapper = null;
            const contextFields = {
                title: {
                    type: 'text',
                    conditionalDisplay: {
                        operator: 'or',
                        fields: [
                            { key:'select', values: '!5' }
                        ]
                    }
                },
                select: {
                    type: 'select',
                    multiple: true
                }
            };

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: [4, 6]
                    }
                }
            });
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: '!5', fieldValue: [4,6], isSingleSelect: false });
            expect(wrapper.findAll(FieldContainer).length).toBe(1);


            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: [5, 7]
                    }
                }
            });
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: '!5', fieldValue: [5,7], isSingleSelect: false });
            expect(wrapper.findAll(FieldContainer).length).toBe(0);
        });

        test('single select (negative)', () => {
            let wrapper = null;
            const contextFields = {
                title: {
                    type: 'text',
                    conditionalDisplay: {
                        operator: 'or',
                        fields: [
                            { key:'select', values: '!6' }
                        ]
                    }
                },
                select: {
                    type: 'select'
                }
            }

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: 3
                    },
                }
            });
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: '!6', fieldValue: 3, isSingleSelect: true });
            expect(wrapper.findAll(FieldContainer).length).toBe(1);

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        select: 6,
                    },
                }
            });
            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: '!6', fieldValue: 6, isSingleSelect: true });
            expect(wrapper.findAll(FieldContainer).length).toBe(0);
        });

        test('or operator', async () => {
            let wrapper = null;
            const contextFields = {
                title: {
                    type: 'text',
                    conditionalDisplay: {
                        operator: 'or',
                        fields: [
                            { key:'check1', values: true },
                            { key:'check2', values: true }
                        ]
                    }
                },
                check1: { type: 'check' },
                check2: { type: 'check' }
            }

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        check1: true,
                        check2: true,
                    },
                }
            });
            expect(wrapper.findAll(FieldContainer).length).toBe(1);

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        check1: false,
                        check2: true,
                    },
                }
            });
            expect(wrapper.findAll(FieldContainer).length).toBe(1);

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        check1: false,
                        check2: false,
                    },
                }
            });
            expect(wrapper.findAll(FieldContainer).length).toBe(0);
        });

        test('and operator', async () => {
            let wrapper = null;
            const contextFields = {
                title: {
                    type: 'text',
                    conditionalDisplay: {
                        operator: 'and',
                        fields: [
                            { key:'check1', values: true },
                            { key:'check2', values: true }
                        ]
                    }
                },
                check1: { type: 'check' },
                check2: { type: 'check' }
            };

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        check1: true,
                        check2: true,
                    },
                }
            })
            expect(wrapper.findAll(FieldContainer).length).toBe(1);

            wrapper = createWrapper({
                propsData: {
                    contextFields,
                    contextData: {
                        title: null,
                        check1: true,
                        check2: false,
                    },
                }
            });
            expect(wrapper.findAll(FieldContainer).length).toBe(0);
        });
    });

    test('expose appropriate props', async () => {
        let wrapper = null;
        const propsData = {
            contextData: {
                title: 'myTitle',
            },
            contextFields: {
                title: {
                    type: 'text',
                    placeholder: 'Title',
                    label: 'Super title label',
                    helpMessage: 'Super help message'
                }
            },
        }

        wrapper = createWrapper({
            propsData,
        });

        expect(wrapper.find(FieldContainer).vm.$attrs).toMatchObject({
            fieldKey: 'title',
            fieldType: 'text',
            fieldProps: {
                type: 'text',
                placeholder: 'Title'
            },
            value: 'myTitle',
            label: 'Super title label',
            helpMessage: 'Super help message'
        });

        wrapper = createWrapper({
            propsData: {
                ...propsData,
                readOnly: true,
            },
        })

        expect(wrapper.find(FieldContainer).vm.$attrs).toMatchObject({
            fieldProps: {
                type: 'text',
                placeholder: 'Title',
                readOnly: true,
            },
        });
    });

    test('call update visibility',  () => {
        let wrapper = null;
        const updateVisibility = jest.fn();
        const contextFields = {
            title: {
                type: 'text',
                conditionalDisplay: {
                    operator: 'or',
                    fields: [{ key:'check', values: true }]
                }
            },
            check: { type: 'check' }
        };

        wrapper = createWrapper({
            propsData: {
                contextFields,
                contextData: {
                    title: 'myTitle',
                    check: true,
                },
                updateVisibility,
            }
        });
        expect(updateVisibility).toHaveBeenCalledTimes(1);
        expect(updateVisibility).toHaveBeenLastCalledWith('title',true);

        wrapper = createWrapper({
            propsData: {
                contextFields,
                contextData: {
                    title: 'myTitle',
                    check: false,
                },
                updateVisibility,
            }
        });
        expect(updateVisibility).toHaveBeenCalledTimes(2);
        expect(updateVisibility).toHaveBeenLastCalledWith('title',false);
    });

    test('handle localized value', async () => {

        function resolveValue({
            type,
            value,
            locale,
            localized = true,
            $form = { localized: true }
        }) {
            const wrapper = createWrapper({
                propsData: {
                    contextData: {
                        title: value,
                    },
                    contextFields: {
                        title: { type, localized },
                    },
                    locale,
                },
                provide: {
                    $form: new Vue({ data:$form }),
                }
            });
            return wrapper.find(FieldContainer).vm.$attrs.value;
        }

        expect(
            resolveValue({
                type: 'text',
                value: null,
            })
        ).toBe(null);

        expect(
            resolveValue({
                type: 'text',
                value: { fr:'texte FR', en:'texte EN' },
                locale: 'fr',
            })
        ).toBe('texte FR');

        expect(
            resolveValue({
                type: 'textarea',
                value: { fr:'texte FR', en:'texte EN' },
                locale: 'fr',
            })
        ).toBe('texte FR');

        expect(
            resolveValue({
                type: 'markdown',
                value: { fr:'texte FR', en:'texte EN' },
                locale: 'fr'
            })
        ).toEqual({ fr:'texte FR', en:'texte EN' });

        expect(
            resolveValue({
                type: 'wysiwyg',
                value: { fr:'texte FR', en:'texte EN' },
                locale: 'fr'
            })
        ).toEqual({ fr:'texte FR', en:'texte EN' });

        expect(
            resolveValue({
                type: 'text',
                value: { fr:'texte FR', en:'texte EN' },
                localized: false,
            })
        ).toEqual({ fr:'texte FR', en:'texte EN' });

        expect(
            resolveValue({
                type: 'text',
                value: { fr:'texte FR', en:'texte EN' },
                $form: { localized: false },
            })
        ).toEqual({ fr:'texte FR', en:'texte EN' });
    });

    xtest('handle localized identifier', () => {
        let wrapper = null;

        wrapper = createWrapper({
            contextData: {
                title: null,
            },
            contextFields: {
                title: { type: 'text', localized: true  },
            },
            locale: 'fr'
        });
        expect(wrapper.find(FieldContainer).vm.$attrs.errorIdentifier).toEqual('title.fr');

        wrapper = createWrapper({
            contextData: {
                title: null,
            },
            contextFields: {
                title: { type: 'text', localized: false  },
            },
            locale: 'fr'
        });
        expect(wrapper.find(FieldContainer).vm.$attrs.errorIdentifier).toEqual('title');
    });
});