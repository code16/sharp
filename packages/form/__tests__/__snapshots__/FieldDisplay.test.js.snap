// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field-display can mount field display 1`] = `<div class="FIELD_CONTAINER_MOCK" fieldkey="title" fieldprops="[object Object]" fieldtype="text" erroridentifier="title" configidentifier="title"></div>`;
